IsekaiMod/Utilities/PatchTools.cs

```diff
@@ -0,0 +1,505 @@
+                return;
+            }
+            if (component == null) { return; }
+            if (component is AddKnownSpell asSpell) {
+                //don't re add spells already added for my class
+                if (asSpell.m_CharacterClass == referenceClass) {
+                    mySpellSet.Add(new SpellReference(asSpell.SpellLevel, asSpell.m_Spell));
+                }
+            }
+            // we do not have a special spell list, so just add all such spells to spells known
+            if (component is AddSpecialSpellList asSpellList && asSpellList.m_CharacterClass == referenceClass) {
+                foreach (var level2 in asSpellList.SpellList.SpellsByLevel) {
+                    foreach (var spell in level2.m_Spells) {
+                        mySpellSet.Add(new SpellReference(level2.SpellLevel, spell));
+                    }
+                }
+            }
+            if (component is AddAbilityUseTrigger trigger && trigger.m_Spellbooks != null && trigger.m_Spellbooks.Length > 0) {
+                trigger.m_Spellbooks = trigger.m_Spellbooks.AddRangeToArray(patchableSpellBooks);
+            }
+            if (component is AddCasterLevelForSpellbook cl && cl.m_Spellbooks != null && cl.m_Spellbooks.Length > 0) {
+                cl.m_Spellbooks = cl.m_Spellbooks.AddRangeToArray(patchableSpellBooks);
+            }
+            if (component is IncreaseSpellSpellbookDC cdc && cdc.m_Spellbooks != null && cdc.m_Spellbooks.Length > 0) {
+                cdc.m_Spellbooks = cdc.m_Spellbooks.AddRangeToArray(patchableSpellBooks);
+            }
+            //check if component is AddFeature
+            if (component is AddFeatureOnClassLevel asFeat) {
+                PatchClassIntoFeatureOfReferenceClass(asFeat.m_Feature.Get(), myClass, referenceClass, mylevel, loopPrevention);
+                //only add our class as an additional class if the original entry was not valid for all classes but was restricted to the correct base class
+                if (asFeat.m_Class != null && asFeat.m_Class == referenceClass
+                    && (asFeat.m_AdditionalClasses == null || !asFeat.m_AdditionalClasses.Contains(myClass))) {
+                    asFeat.m_AdditionalClasses.AddItem(myClass);
+                }
+            }
+            if (component is MonkNoArmorFeatureUnlock addUnarmedFact) {
+                var fact = addUnarmedFact.m_NewFact.Get();
+                if (fact != null) {
+                    if (fact is BlueprintFeature feature2) {
+                        PatchClassIntoFeatureOfReferenceClass(feature2, myClass, referenceClass, mylevel, loopPrevention);
+                    }
+                    if (fact is BlueprintProgression progression2) {
+                        PatchClassIntoFeatureOfReferenceClass(progression2, myClass, referenceClass, mylevel, loopPrevention);
+                    }
```

Comment:
@kjk001 Not sure if this is intended but `BlueprintProgression` is a subclass of `BlueprintFeature` so `PatchClassIntoFeatureOfReferenceClass` would be redunantly called twice if `fact` is a `BlueprintProgression`.

Reply:
Also `BlueprintFeature` is a subclass of `BlueprintUnitFact` so `HandleComponent` will be called for a `BlueprintFeature` and `BlueprintProgression`